### ISSUE TITLE
Run focused scenario

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber/vscode#69]
### 🤔 What's the problem you're trying to solve?

I don't want to leave my editor to run a scenario

### ✨ What's your proposed solution?

A keyboard shortcut for running the scenario on the current line.

This would run Cucumber with the `message` formatter, send the messages to a temporary file, parse the contents of that file (using [this class](https://github.com/cucumber/message-streams/blob/main/src/NdjsonToMessageStream.ts)) and decorate lines in the code with a background indicating the status. Errors can be expanded straight in the editor. Stack trace elements would be links.

How to execute Cucumber depends on the Cucumber implementation, so the editor would have to know how to deal with each implementation. We should provide a setting to allow users to override how Cucumber is run.

**Comment by gasparnagy** [comment]
I have just payed with the https://marketplace.visualstudio.com/items?itemName=robocorp.robotframework-lsp extension for Robot Framework. They integrate with the tests tab of vscode... pretty nice.

![image](https://user-images.githubusercontent.com/144701/174065410-35d571a3-ce75-47e3-a66f-5846d21beefd.png)

**Comment by sijakret** [comment]
I was just wondering about a feature like this as well.
It would increase ergonomics quite a bit!

Could be implemented very similarly to how npm scripts are made runnable from the code
<img width="200" alt="image" src="https://user-images.githubusercontent.com/16835620/216594845-d2820f0a-b0fa-4879-ae08-2aa47fdae48e.png">

So when you hover over one of 

* feature
* scenario
* tag

It could show a direct way to execute the corresponding test(s).
For instance via a configurable commandline or vscode command.
Is something like this on the roadmap?

**Comment by yaffol** [comment]
This feature would be huge for me. I really want to write my tests as executable specifications, and we are doing that on the backend, but I'm not yet happy with the tooling for the frontend.
I am using VSCode and playwright for most of the frontend e2e tests, and I don't yet find a combination of toolchains that works ergonomically enough to honestly recommend it to other members of the team, so I don't yet adopt BDD on the frontend.

Something really nice about Cypress was the DX of the cy:open Watcher - and the VSCode test explorer is close to that. If this (test explorer integration) were working with the cucumber-vscode extension, I would feel I could hand the integration over to the rest of the team as it gives clear visibility of the tests.

I realise you can always see test output by dropping to the terminal, but there is strong ergonomics behind not having to leave you IDE, especially for frontend development, in my experience.

**Comment by Artalus** [comment]
+1 for a better integration of Feature files with VS Code Test Explorer capabilities. Would love to see the list of scenarios in the side tab, as well as to have a neat `Run Test | Debug Test` lens-button above the scenario name...

**Comment by nextlevelbeard** [comment]
+1 Would also love to see being able to define a custom command in the extension settings 
Useful for using Cucumber with other frameworks like [webdriverIO](https://github.com/webdriverio/webdriverio)

**Comment by kieran-ryan** [comment]
@Artalus and @nextlevelbeard, appreciate the bump on this - improves insight into the desire for this feature - would be a significant enhancement. @nextlevelbeard if you could also pop a thumbs up (👍) on the top-level comment that would be great.

We are quite reliant on community contributions at present, which can be a challenge with larger features such as this. Ways to help can include signposting relevant implementation details, linking other extensions matching elements of your desired functionality; or of course, opening a pull request (even if only contains small or partial implementation details) - which can greatly increase the chances of this work being picked up, and that we could support on.

Thanks for your interest and engagement in the project - hope you can continue to raise these points, it is very much valued.

**Comment by 68267a** [comment]
I'm new to both Cucumber AND Java, so I'm not sure what I'm asking. 

I was messing with the 10-minute tutorial and I noticed that IntelliJ can run and debug each scenario individually. It feels like VS Code should be able to do that as well, but I don't know how to set it up. See screenshot. 

Is this the same topic/issue/conversation? Thanks

![image-1](https://github.com/cucumber/vscode/assets/33899447/ef5acc15-23fa-442e-a608-4fe2b7219c0a)

**Comment by kieran-ryan** [comment]
> I was messing with the 10-minute tutorial and I noticed that IntelliJ can run and debug each scenario individually. It feels like VS Code should be able to do that as well, but I don't know how to set it up. See screenshot.
> 
> Is this the same topic/issue/conversation? Thanks

Yep, you've absolutely got the right issue for this @68267a. Feel free to thumbs-up the top issue as well to track interest 👍

**Comment by JDUNNIN** [comment]
Sharing the following in case it helps out:

I found for Cucumber (for Java) that opening a Java class which uses `org.junit.runner.RunWith` and `io.cucumber.junit.CucumberOptions`, then means on the left side (where you'd put breakpoints) there is a small `Run Test` and `Debug Test` button alongside the `public class` line. Not sure if this is native VSCode or if it was part of another extension called Cucumber (Gherkin) Full Support I use. Example class I used:
```
  @RunWith(Cucumber.class)
  @CucumberOptions(
      features = "src/test/resources/features",
      tags = "@MyTagToTest",
      glue = {"com.example.project"},
      strict = true,
      snippets = SnippetType.CAMELCASE
  )
public class MyCucumberTestRunner {
    // No code needed here
}
```
